smolyay/surrogate.py

```diff
@@ -1,546 +1,828 @@
+
+    @property
+    def regularization(self):
+        """:class:RegularizationHelper: constant of the L2 and L1 term"""
+        return self._regularization
+
+    @regularization.setter
+    def regularization(self, value):
+        if not value is None:
+            if not isinstance(value, RegularizationHelper):
+                raise ValueError("Regression must be a RegularizationHelper")
+        if self.regularization != value:
+            self._regularization = value
+            self._valid_cache = False
+
+    @property
+    def basis_sets(self):
+        """list of BasisFunctionSet: the set of basis functions for the terms."""
+        return self._basis_sets
+
+    @property
+    def coefficients(self):
+        """numpy.ndarray: the coefficients of the terms"""
+        return self._coefficients
+
+    def predict(self, X):
         """Evaluate surrogate at a given input.
 
-        Parameter
-        ---------
-        x: list (for dimension > 1), or float (for dimension = 1)
-            Input.
+        Parameters
+        ----------
+        X: array-like with shape (n_samples, num_dimensions) or list of object
+            Points at which the model is evaluated
 
         Returns
         -------
-        float
+        ndarray of shape (n_samples,) or (n_samples, n_features)
             Surrogate output at x.
 
         Raises
         ------
-        ValueError
+        RuntimeError
             For surrogate to be evaluated, function needs to be trained.
-        IndexError
-            Input of the surrogate must be of length of the
-            function's dimensionality.
         ValueError
             Input must lie in domain of surrogate.
+        NotImplementedError
+            Predict after fitting to gradient not supported.
         """
-        if self._coefficients is None:
-            raise ValueError('Function needs training!')
-        input_shape = numpy.shape(x)
-        x = numpy.array(x, copy=False, ndmin=2)
-        if self.dimension == 1:
-            if len(input_shape) <= 1:
-                # the cast to 2d puts these in wrong order, so transpose
-                x = x.T
-            elif x.shape[-1] > 1:
-                x = x[..., numpy.newaxis]
-        
-        if x.shape[-1] != self.dimension:
-            raise IndexError("Input must match dimension of domain")
-        if self.dimension > 1:
-            oob = any(numpy.any(x[..., i] < self.domain[i][0]) or 
-                    numpy.any(x[..., i] > self.domain[i][1]) 
-                    for i in range(self.dimension))
-        else:
-            oob = (numpy.any(x < self.domain[0]) or 
-                    numpy.any(x > self.domain[1]))
+        # validate inputs
+        X = numpy.array(X, ndmin=2)
+        if X.shape[1] != self.num_dimensions:
+            raise IndexError("Must be 2D array with shape (n_samples, n_features)")
+        if not self._valid_cache:
+            raise RuntimeError("Model must be fit!")
```

Comment:
Put this first in the predict method, nothing should be done if this is not valid.